docs/HowToFAQ/Errors-Compiler.md

```diff
@@ -1,104 +1,529 @@
+there may be no program statements that have an arbitrary, even if deterministic, result.
+The [`if` with a binding guard](../DafnyRef/DafnyRef#sec-binding-guards) (`if ... :| `) checks that the 
+given predicate is satisfiable by some value. If not, then the 'else' branch is executed;
+but if so, the 'then' branch is executed with an arbitrary value that satisifies the predicate.
+Because of this arbitrary selection, the if-with-binding-guard is not permitted with `--enforce-determinism`,
+even if there is exactly one value that satisfies the predicate.
+
+## **Error: case-based if statement forbidden by the --enforce-determinism option**
+
+<!-- %check-run %options --enforce-determinism -->
+```dafny
+method m(k: int) {
+  var i := k;
+  if
+  {
+    case i >= 0 => i := i - 1;
+    case i <= 0 => i := i + 1;
+  }
+}
+```
+
+The `--enforce-determinism` option requires a target program to be deterministic and predictable:
+there may be no program statements that have an arbitrary, even if deterministic, result.
+The case-based if statement allows the case guards to be evaluated in an arbitrary order and
+an arbitrary one of those found to be true is chosen to be executed.
+Hence the case-based if is not permitted with `--enforce-determinism`.
+
+To enforce a deterministic order to the evaluation, use a chain of if-then-else statements.
+
+
+## **Error: nondeterministic loop forbidden by the --enforce-determinism option**
+
+<!-- %check-run %options --enforce-determinism -->
+```dafny
+method m(b: bool) decreases * {
+  while * 
+    decreases *
+  { }
+}
+```
+
+The `--enforce-determinism` option requires a target program to be deterministic and predictable:
+there may be no program statements that have an arbitrary, even if deterministic, result.
+Hence this 'non-deterministic while' (`while *`) statement is not permitted with `--enforce-determinism`.
+
+## **Error: case-based loop forbidden by the --enforce-determinism option**
+
+<!-- %check-run %options --enforce-determinism -->
+```dafny
+method m(k: int) {
+  var i := k;
+  while 
+    decreases if i < 0 then -i else i
+  {
+    case i > 0 => i := i - 1;
+    case i < 0 => i := i + 1;
+  }
+}
+```
+
+The `--enforce-determinism` option requires a target program to be deterministic and predictable:
+there may be no program statements that have an arbitrary, even if deterministic, result.
+The case-based while statement allows the case guards to be evaluated in an arbitrary order and
+an arbitrary one of those found to be true is chosen to be executed.
+Hence the case-based loop is not permitted with `--enforce-determinism`.
+
+To enforce a deterministic order to the evaluation, use a a chain of if-then-else statements
+or series of `if` statements in which the then-branch ends in a continue statement.
+
+## **Error: compiler currently does not support assignments to more-than-6-dimensional arrays in forall statements**
```

Comment:
6 dimensions ought to be enough for anybody. ;)

Reply:
And you can do more than 6 if the indices have obvious ranges.

Reply:
I haven't come up with a failing example yet.